flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/TimeBoundedStreamJoinOperator.java

```diff
@@ -0,0 +1,398 @@
+			timeServiceManager.advanceWatermark(mark);
+		}
+
+		output.emitWatermark(new Watermark(mark.getTimestamp() - getWatermarkDelay()));
+	}
+
+	private void collect(T1 left, T2 right, long leftTs, long rightTs) throws Exception {
+		collector.setAbsoluteTimestamp(leftTs);
+		context.leftTs = leftTs;
+		context.rightTs = rightTs;
+		userFunction.processElement(left, right, context, this.collector);
+	}
+
+	private void removeFromLhsUntil(long maxCleanup) throws Exception {
+
+		// setup state
+		if (this.lastCleanupRightBuffer.value() == null) {
+			this.lastCleanupRightBuffer.update(0L);
+		}
+
+		// remove elements from leftBuffer in range [lastValue, maxCleanup]
+		Long lastCleanUpRight = lastCleanupRightBuffer.value();
+		for (long i = lastCleanUpRight; i <= maxCleanup; i++) {
+			leftBuffer.remove(i);
+		}
+
+		lastCleanupRightBuffer.update(maxCleanup);
+	}
+
+	private void removeFromRhsUntil(long maxCleanup) throws Exception {
+
+		// setup state
+		if (this.lastCleanupLeftBuffer.value() == null) {
+			this.lastCleanupLeftBuffer.update(0L);
+		}
+
+		// remove elements from rightBuffer in range [lastValue, maxCleanup]
+		Long lastCleanupLeft = lastCleanupLeftBuffer.value();
+		for (long i = lastCleanupLeft; i <= maxCleanup; i++) {
+			rightBuffer.remove(i);
+		}
+
+		lastCleanupLeftBuffer.update(maxCleanup);
+	}
+
+	private boolean shouldBeJoined(long leftTs, long rightTs) {
+		long elemLowerBound = leftTs + lowerBound;
+		long elemUpperBound = leftTs + upperBound;
+
+		boolean lowerBoundOk = (lowerBoundInclusive)
+			? (elemLowerBound <= rightTs)
+			: (elemLowerBound < rightTs);
+
+		boolean upperBoundOk = (upperBoundInclusive)
+			? (rightTs <= elemUpperBound)
+			: (rightTs < elemUpperBound);
+
+		return lowerBoundOk && upperBoundOk;
+	}
+
+	private void addToLeftBuffer(T1 value, long ts) throws Exception {
+		long bucket = calculateBucket(ts);
+		Tuple3<T1, Long, Boolean> elem = Tuple3.of(
+			value, // actual value
+			ts,    // actual timestamp
+			false  // has been joined
+		);
+
+		List<Tuple3<T1, Long, Boolean>> elemsInBucket = leftBuffer.get(bucket);
+		if (elemsInBucket == null) {
+			elemsInBucket = new ArrayList<>();
+		}
+		elemsInBucket.add(elem);
+		leftBuffer.put(bucket, elemsInBucket);
+	}
+
+	private void addToRightBuffer(T2 value, long ts) throws Exception {
+		long bucket = calculateBucket(ts);
+		Tuple3<T2, Long, Boolean> elem = Tuple3.of(
+			value, // actual value
+			ts,    // actual timestamp
+			false  // has been joined
+		);
+
+		List<Tuple3<T2, Long, Boolean>> elemsInBucket = rightBuffer.get(bucket);
+		if (elemsInBucket == null) {
+			elemsInBucket = new ArrayList<>();
+		}
+		elemsInBucket.add(elem);
+		rightBuffer.put(bucket, elemsInBucket);
+	}
+
+	private long calculateBucket(long ts) {
+		return Math.floorDiv(ts, bucketGranularity);
+	}
+
+	public long getWatermarkDelay() {
+		// TODO: Adapt this to when we use the rightBuffer or min timestamp
+		return (upperBound < 0) ? 0 : upperBound;
```

Comment:
Watermark delay should be max(abs(lower_bond), abs(upper_bond)), because both left and right may output watermark to join.

Reply:
I added some sketches to the design document about my thoughts behind this, maybe continue our discussion there?